HybridCar.hpp

```diff
@@ -1,19 +1,23 @@
+#pragma once
+
+#include "ElectricCar.hpp"
 #include "ElectricEngine.hpp"
+#include "PetrolCar.hpp"
 #include "PetrolEngine.hpp"
 
-class HybridCar
+class HybridCar : public PetrolCar, public ElectricCar
 {
 public:
-    HybridCar(PetrolEngine* petrolEng, ElectricEngine* electricEng);
-    ~HybridCar();
-    void turnLeft();
-    void turnRight();
-    void brake();
-    void accelerate(int speed);
-    void charge();
-    void refuel();
+    HybridCar(PetrolEngine* petrolEngine, ElectricEngine* electricEngine);
+    virtual ~HybridCar();
+
+    void restore() override;
+    void changeGear(int gear) override;
+    virtual void setEngine(PetrolEngine*) override;     // overloaded (not used)
+    virtual void setEngine(ElectricEngine*) override;   // overloaded (not used)
+    virtual void setEngine(PetrolEngine*, ElectricEngine*) override;
 
+private:
     PetrolEngine* petrolEngine_;
     ElectricEngine* electricEngine_;
```

Comment:
No, this is problematic. You have 4 engines in your HybridCar. 2 are here, 1 is in a base class PetrolCar and 1 in a base class ElectricCar. HybridCar class should have no engine.

Reply:
I don't get it. If removed there's an error: class ‘HybridCar’ does not have any field named ‘petrolEngine_’ ... 
‘electricEngine_’. It's only a pointer - not new objects have been created yet. In other people's repos this solution was approved even with memory leaks (there are raw or smart pointers)!  I have checked. I've got the same memory leaks as them (56 bytes).

Reply:
which one? probably I have overlooked.
You can make members in base classes protected and you will have access to them without having additional pointers here.

Reply:
I have tried even with public access before. I must refactor the whole code. Not an easy life.

Reply:
Indeed, there were too many engines. I misused those pointers. It was enough to call the constructors of petrol & electric cars when creating hybrid car object. So obvious now! Moreover other appropriate methods were reused form those classes to change the engine and restore (refuel). Thank you for your deep and accurate insight!